### ISSUE TITLE
clear() method ignores tags, clears entire store's cache

### ISSUE THREAD

**Issue by zlanich** [spatie/laravel-responsecache#331]
Inside your `ResponseCacheRepository::clear()` method, the first call to `$this->cache->clear()` needs to be `$this->cache->flush()`. If you call `TaggableCache::clear()`, it delegates to the Store's flush method, which flushes the entire store. If you call `TaggableCache::flush()`, this method gets overridden by `TaggableCache` implementations like `RedisTaggedCache`, which appropriately deletes values by segment (tag, ref), etc.

Your `$this->cache->tags(config('responsecache.cache_tag'))->flush();` call is already doing this correctly. Looks like a simple typo, but it's wreaking havoc in our prod app. For now, I overrode a bunch of classes to fix it myself.

```php
public function clear(): void
    {
        if ($this->isTagged($this->cache)) {
            $this->cache->flush(); // Correct
            // $this->cache->clear(); // Incorrect

            return;
        }

        if (empty(config('responsecache.cache_tag'))) {
            $this->cache->clear();

            return;
        }

        $this->cache->tags(config('responsecache.cache_tag'))->flush();
    }
```

**Comment by freekmurze** [comment]
Hi, could you please create a PR for this? Make sure to include tests to make sure that everything is working correctly.

**Comment by zlanich** [comment]
I am willing to do so, but won’t have time for a while. I’m not yet sure
how to write unit tests for something like tagged cache, since not all
drivers support it. I will look into this asap, but no guarantee on when.

<https://www.facebook.com/zachlanich>
Zach Lanich

Owner/CTO | weCreate - Website Design & Marketing

phone:  814.823.3602
site:  www.wecreate.com/
<https://www.facebook.com/zachlanich>
<https://www.linkedin.com/in/zachlanich/>
<https://www.facebook.com/wecreatellc/>

On January 1, 2021 at 4:35:55 PM, Freek Van der Herten (
notifications@github.com) wrote:

Hi, could you please create a PR for this? Make sure to include tests to
make sure that everything is working correctly.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/spatie/laravel-responsecache/issues/331#issuecomment-753385539>,
or unsubscribe

**Comment by merlinblack** [comment]
This bug makes response cache unusable for us. Imagine our surprise after a lot of debugging, finding that ResponseCache::clear() deletes Laravel's worker queues, and everything else in Redis.

**Comment by zlanich** [comment]
Omg, I never thought about the fact that it would clear Redis queues if
you’re not using separate databases. That’s awful. For us, it was
destroying our production caches, effectively making all of our performance
optimizations moot. Obviously not intentional, but a definite major
oversight by accident. Worth fixing right away.


Best Regards,

Zach Lanich
*Business Owner, Entrepreneur, Creative*
*Owner/CTO*
weCreate LLC
*www.WeCreate.com <http://www.WeCreate.com>*

From: Nigel Atkinson <notifications@github.com> <notifications@github.com>
Reply: spatie/laravel-responsecache
<reply@reply.github.com>
<reply@reply.github.com>
Date: January 27, 2021 at 10:26:26 PM
To: spatie/laravel-responsecache <laravel-responsecache@noreply.github.com>
<laravel-responsecache@noreply.github.com>
CC: zlanich <zach@zachlanich.com> <zach@zachlanich.com>, Author
<author@noreply.github.com> <author@noreply.github.com>
Subject:  Re: [spatie/laravel-responsecache] clear() method ignores tags,
clears entire store's cache (#331)

This bug makes response cache unusable for us. Imagine our surprise after a

**Comment by Kovah** [comment]
Any news on this?
I actually thought that I became insane because after flushing some caches, entirely unrelated tags got deleted too. No wonder: I called ResponseCache::clear() before.

I would submit a PR, but have no idea how to write the tests for these use cases.

**Comment by zlanich** [comment]
@Kovah Though I do not personally have the time to pull this down and write tests for it, I can confirm that my fix is 100% working in production for us. We actually use the caching system for temporarily storing other data as well, and many of our features wouldn't work at all without this fix. The issue/fix is actually pretty obvious when using XDebug and actually finding your way into these lower PHP files. It's just an oversight in regards to the Redis driver class(es).

The easiest way to test this without an actual unit test is to go into Tinker and do something like this:
```php
// Set a test cache item
\Cache::tags(['dont-clear-me'])->put('test', 123, null);
// Read it back to make sure it's there
\Cache::tags(['dont-clear-me'])->get('test');
// Clear Response Cache (which would normally flush entire store w/o my fix)
\ResponseCache::clear();
// Read it back to make sure it's STILL there
\Cache::tags(['dont-clear-me'])->get('test');
```

If you test that with and without my fix, you should find that the test key either sticks or gets flushed. Please let me know if this is the case for you!

**Comment by merlinblack** [comment]
Likewise - we are using this fix in production also. Without it, we would have had to ditch is library altogether. The bug really is a show stopper. I did some testing in tinker similar to @zlanich above and also confirmed using a Redis viewer.

**Comment by zlanich** [comment]
@merlinblack Thanks a lot for verifying. I'm super happy it's working for people. If I wasn't working 14hrs/day right now. I would have pulled it down already and submitted a PR. Hopefully I or someone else will have time to soon.

**Comment by Kovah** [comment]
I deployed my project to production with the hotfix yesterday and everything seems to be working as expected. Thank you for your help @zlanich!

**Comment by trevorpe** [comment]
@Kovah Considering using this library, so I figured I'd so some digging on this issue and I found the reason for the failing tests.

In `ResponseCacheRepository`, the `tags()` method builds a new instance `ResponseCacheRepository` with a new `TaggedCache`.
```php
    public function tags(array $tags): self
    {
        return new self($this->responseSerializer, $this->cache->tags($tags));
    }
```

The problem comes from the `$this->cache->tags($tags)` call. This new `TaggedCache` only has the tag provided by the route middleware ("foo" in this case), and not the global tag provided in the config ("tagging-test"). Calling `tags()` on a `TaggedCache` instance does not concatenate the array of tags. (This is presumably a Laravel design decision or bug.)

So when the the response cache is cleared using `$this->app['responsecache']->clear()`, only entries tagged with "tagging-test" are cleared. Since the cache that stored the responses only had the "foo" tag, that statement doesn't clear the cached response and the cached response gets returned again, failing the test.

Not sure what the desired behaviour is here. I would assume calling `$this->app['responsecache']->clear()` would clear all cached responses, which means if there's a global tag, all cached responses must have it. If that's the case, the following implementation for `ResponseCacheRepository::tags()` fixes it with nesting behaviour:
```php
    public function tags(array $tags): self
    {
        if ($this->isTagged($this->cache)) {
            $tags = array_merge($this->cache->getTags()->getNames(), $tags);
        }
        return new self($this->responseSerializer, $this->cache->tags($tags));
    }
```

An alternative is to only add the global tag and not a nesting behaviour.

**Comment by Kovah** [comment]
@freekmurze if you approve this change I would update my pull request accordingly. Not sure if you want the package to behave like this, but makes sense to me to flush both the configuration tags and the route tags if you use the `ResponseCache::clear();` method.

**Comment by freekmurze** [comment]
> makes sense to me to flush both the configuration tags and the route tags if you use the ResponseCache::clear(); method.

That sound good! Include that in the PR.

**Comment by dimti** [comment]
Not for the PR.
Laravel 6.x https://github.com/dimti/laravel-responsecache/tree/redis-tagged-cache-flush-for-laravel-6

**Comment by volkv** [comment]
Can confirm such unwanted behaviour. More than that `ResponseCache::clear()` clears not only its own whole store cache but ALL existing stores caches

**Comment by freekmurze** [comment]
I've merged in a fix for this (#344)
